src/test/scala/sigmastate/utils/ByteReaderWriterImpSpecification.scala

```diff
@@ -1,8 +1,53 @@
 package sigmastate.utils
 
-import org.scalatest.{Matchers, PropSpec}
+import org.scalacheck.{Arbitrary, Gen}
 import org.scalatest.prop.PropertyChecks
+import org.scalatest.{Matchers, PropSpec}
+import sigmastate.serialization.Serializer
+import sigmastate.serialization.generators.ValueGenerators
+
+class ByteReaderWriterImpSpecification extends PropSpec
+  with ValueGenerators
+  with PropertyChecks
+  with Matchers {
+
+  private val seqPrimValGen: Gen[Seq[Any]] = for {
+    length <- Gen.chooseNum(1, 100)
+    anyValSeq <- Gen.listOfN(length,
+      Gen.oneOf(
+        Arbitrary.arbByte.arbitrary,
+        Arbitrary.arbShort.arbitrary,
+        Arbitrary.arbInt.arbitrary,
+        Arbitrary.arbLong.arbitrary,
+        arrayGen[Byte]))
+  } yield anyValSeq
+
+  private implicit val arbSeqPrimVal: Arbitrary[Seq[Any]] = Arbitrary(seqPrimValGen)
 
-class ByteReaderWriterImpSpecification extends PropSpec with PropertyChecks with Matchers {
+  property("round trip serialization/deserialization of arbitrary value list") {
+    forAll { values: Seq[Any] =>
+      val writer = values.fold(Serializer.startWriter()) { case (w: ByteArrayWriter, anyVal: Any) =>
```

Comment:
@greenhat There are some issues, mostly related to Scala style
1) AnyVal is a type in Scala, so identifier anyVal suggests it has AnyVal type (not Any)
2) Over functionalization is extreme style not encouraged in general. 
Especially is the context of mutable API. 
Writer/Reader is mutable API, in which sequence of operation is primary and everything else is secondary.
Functional operations works best, with immutable data-centric APIs.
I suggest to use imperative `for (anyVal <- values) {}` which make sequence of operations clear.

Reply:
@ergomorphic Agreed.
I got carried away with this `fold`. Although I did see the mutable API I just wanted to encode the notion of using the returned writer which does not make much sense.

Reply:
@ergomorphic BTW, would `foreach` be a better option than `for` here? I found it more concise with less indenting.
```
      for(any <- values) {
        any match {
          case ...
        }
      }
```
vs 
```
      values.foreach {
        case ...
      }
```